### PR TITLE
Remove the Hyrax::SearchBuilder

### PR DESCRIPTION
**Pull request by jcoyne**
This class is duplicative of what is already in
`app/models/search_builder.rb` so it only offers unneeded rediretion.

Also moved search bulders into the `Hyrax::My` namespace for those that
only show objects deposited by the current user.

### CODE REVIEW

**Comment by atz** [comment]
Wasn't the point to have a common superclass of all hyrax search builders?

**Comment by atz** [comment]
see: https://github.com/projecthydra-labs/hyrax/issues/707#issuecomment-292996882

**Comment by jcoyne** [comment]
@atz the reasoning here is that people may want to install blacklight-range-limit or other blacklight plugins which modify `::SearchBuilder`.  I *think* this is better from that regard.  @jrochkind what do you think?

File: app/search_builders/hyrax/search_builder.rb [hunk 1]
```diff
@@ -1,15 +1,8 @@
-# TODO: make this a mixin and generate it into ::SearchBuilder
 module Hyrax
-  class SearchBuilder < Blacklight::SearchBuilder
```
Comment (atz) [inline app/search_builders/hyrax/search_builder.rb]: 👍

**Comment by atz** [comment]
I agree `::SearchBuilder` is blacklight's conventional target for modification.  I just want to minimize the number of generators/destroyers we have all trying to kludge our way into the same app class that itself isn't versioned or controlled.  Ideally `Hyrax::SearchBuilder` inherits from `::SearchBuilder`, as you have here.

**Comment by jrochkind** [comment]
@jcoyne: I can't completely follow all the implications, I gotta admin. But for starters, def all the SearchBuilders should have parallel inheritance hiearchies. I _think_ this accomplishes that? If so, that's good. 

Additionally there should be clear ways to customize search builders. And _probably_ clear ways to customize _all_ search builders. 

I think this makes all of them inherit from app local `::SearchBuilder` so that customizations there (whether manual and local, or injected by plugins like range limit) will apply to _all_ search builders. This strikes me as an improvement. (But are there any cases where someone is going to accidentally effect all search builders, when they don't mean to?  I don't _think_ there are any big gotchas here, but I don't feel like my mind is completely around the existing architecture and all those dozens of search builders and what they do). 

And I _think_ you have tried to provide an easier way to customize individual search builders instead of all of them too? Maybe with the use of the `builder = search_builder(controller)` formulation?  I think that is a use case that needs to have a path; I don't understand enough of the architecture to know if this is the optimal one, or how it would be intended to be used. Some docs with examples of "effect just specific search builder" use cases is needed, somewhere. 

So I am cautiously optimistic about this change, although I don't understand it enough to be sure. 

There is existing documentation that will have to be changed, I think. And ideally new docs that explain the use cases for "effect all search builders with custom logic" and "effect just certain search builders with custom logic", as well as ideally providing context for why there are so many search builders and what that architecture is about.

**Comment by jrochkind** [comment]
@atz 

>  but I don't understand that we have changed our reasoning from when we originally made Hyrax::SearchBuilder.

If anyone could share what that reasoning was, it would help me understand what's up here. Reasoning like that is a great thing to capture in class head rdoc, for future readers. The code alone can't show reasoning, motivation, or intent.

**Comment by jcoyne** [comment]
Here is the original commit for `Sufia::SearchBuilder` (later renamed `Hyrax::SearchBuilder`)  https://github.com/projecthydra/sufia/commit/0c22d71198f564fa653166b932c8ffac043d914c It appears that it was created at the time because we only had one search builder for everything.  Now that it's been well decomposed, I don't think it serves any utility.

**Comment by jrochkind** [comment]
It _might_ make sense to have all the `Hyrax::My` search builders inherit from a common `My` superclass -- which still inherits from `::SearchBuilder` for cases where someone wants to add logic to all `My` search builders, but _not_ to other search builders.  I don't think there's any good way to add logic just to the `MySearchBuilderBehavior` module.  In fact, if that's true, and a common `My` superclass is useful, perhaps it should just contain `MySearchBuilderBehavior` behavior directly, no need for a module concern. (I consider mixins to be effectively multiple inheritance, and super useful in some cases, but not to be used gratuiously). 

But I don't totally understand if there is a use case for "add custom logic to all `My` search builders only".  I don't entirely understand why there are dozens of search builders in the first place, or what each of them is used for. 

I think ideally we'd begin with developer use cases here, which the architecture could be tested against, and could also form the nucleous of some docs with examples. 

I know I have a use case for blacklight_range_limit, but I don't totally understand if I want/need that injected into _all_ the search builders, which search builders I'd need it injected into, or if there are any it _shouldn't_ be injected into (as opposed to not neccesary but harmless to do so).

**Comment by atz** [comment]
@jrochkind 

> But I don't totally understand if there is a use case for "add custom logic to all My search builders only". I don't entirely understand why there are dozens of search builders in the first place, or what each of them is used for.


I imagine that might be for some custom permissions or admin capability.

@jcoyne 

> Here is the original commit for Sufia::SearchBuilder (later renamed Hyrax::SearchBuilder) projecthydra/sufia@0c22d71 It appears that it was created at the time because we only had one search builder for everything. Now that it's been well decomposed, I don't think it serves any utility.


I appreciate the context, but that doesn't match my understanding of the intent here, long after that commit in Sufia.  One of the [lines you remove](https://github.com/projecthydra-labs/hyrax/blame/master/app/search_builders/hyrax/search_builder.rb#L1) said:

```
# TODO: make this a mixin and generate it into ::SearchBuilder
```

But this PR doesn't accomplish that, right?   I guess I'm ok w/ inheriting from a `::SearchBuilder` that doesn't exist until after generation, but it still feels weird.  Unblocking: if somebody else wants to approve it, fine.

**Comment by jrochkind** [comment]
@atz Maybe I'm misunderstanding, but the point of inheriting from `::SearchBuilder` that doesn't exist until after generation, is that the local `::SearchBuilder` is the place where the user (or other plugins) can add code meant to apply to _all_ app search builders.  Right, are you with me? 

I'm not sure if you're thinking there's no need to do this (or that it may even be inappropriate), or if you are considering a different way to do this, or something else entirely. 

One way or another, I think a solution is needed to the issues raised in #707. I _think_ maybe this does that -- at least I believe it's intended to?  If you have a different approach in mind though, I'd be definitely interested in hearing it!

**Comment by jrochkind** [comment]
I _think_ I like this, and I think it provides an architecture that answers some of the questions/problems in #707. 

But I'm reluctant to be the approver myself:

* I don't totally understand everything in here, it's a large refactor
* I don't totally understand the use cases the original (before this PR) architecture was meant to be solving, why it was like that in the first place, and if this PR breaks any of them. 
* I am not sure if the backwards incompat implications of this PR, and if they are significant, if there's another way to approach #707 with fewer. 

I think ideally a (potentially wider) discussion with words not just code would happen here about the use cases intended to be supported by the way hyrax uses SearchBuilders, and potential backwards compat issues of this change, if any.  But if some other committer wants to approve it as is, I'm not blocking!  Perhaps other people understand this context better than me, I'm just not comfortable approving/merging with my understanding as it is.

**Comment by jrochkind** [comment]
>> But I don't totally understand if there is a use case for "add custom logic to all My search builders only". I don't entirely understand why there are dozens of search builders in the first place, or what each of them is used for.

> I imagine that might be for some custom permissions or admin capability.

I think that suggests giving all the `My` searchbuilders a common superclass, which itself `< ::SearchBuilder`, so that common superclass is a place to inject behavior you want _only_ for `My` builders, and not for everything. What do you think, @jcoyne ?

**Comment by mjgiarlo** [comment]
@jcoyne :point_up:

**Comment by jcoyne** [comment]
@mjgiarlo @jrochkind isn't that what ` Hyrax::My::SearchBuilder` is? A common search builder for the `My*` builders?

**Comment by jrochkind** [comment]
@jcoyne you are correct, I missed that, sorry!

**Comment by jeremyf** [comment]
@jcoyne Is this something you are still interested in?

**Comment by jcoyne** [comment]
@jeremyf Yes, it got stale waiting for an approval.

**Comment by jeremyf** [comment]
@jcoyne I'll work on the merge conflicts today

**Pull request closed without merging**